ledger/src/executor.rs

```diff
@@ -0,0 +1,421 @@
+        borrow::Cow,
+        collections::{HashMap, HashSet},
+        result,
+        sync::{Arc, RwLock},
+        thread::{self, Builder, JoinHandle},
+    },
+};
+
+pub(crate) struct TransactionBatchWithIndexes<'a, 'b> {
+    pub batch: TransactionBatch<'a, 'b>,
+    pub transaction_indexes: Vec<usize>,
+}
+
+/// Callback for accessing bank state while processing the blockstore
+pub type ProcessCallback = Arc<dyn Fn(&Bank) + Sync + Send>;
+
+// get_max_thread_count to match number of threads in the old code.
+// see: https://github.com/solana-labs/solana/pull/24853
+lazy_static! {
+    static ref PAR_THREAD_POOL: ThreadPool = rayon::ThreadPoolBuilder::new()
+        .num_threads(get_max_thread_count())
+        .thread_name(|ix| format!("solBstoreProc{:02}", ix))
+        .build()
+        .unwrap();
+}
+
+const DEFAULT_CONFLICT_SET_SIZE: usize = 30;
+
+pub struct ReplayResponse {
+    pub result: Result<()>,
+    pub timing: ExecuteTimings,
+    pub batch_idx: Option<usize>,
+}
+
+/// Request for replay, sends responses back on this channel
+pub struct ReplayRequest {
+    pub tx_and_idx: (SanitizedTransaction, usize),
+    pub bank: Arc<Bank>,
+    pub transaction_status_sender: Option<TransactionStatusSender>,
+    pub replay_vote_sender: Option<ReplayVoteSender>,
+    pub cost_capacity_meter: Arc<RwLock<BlockCostCapacityMeter>>,
+    pub tx_cost: u64,
+    pub log_messages_bytes_limit: Option<usize>,
+    pub entry_callback: Option<ProcessCallback>,
+    pub batch_idx: Option<usize>,
+}
+
+pub struct Replayer {
+    threads: Vec<JoinHandle<()>>,
+    request_sender: Sender<(Sender<ReplayResponse>, ReplayRequest)>,
+}
+
+pub struct ReplayerHandle {
+    request_sender: Sender<(Sender<ReplayResponse>, ReplayRequest)>,
+    response_sender: Sender<ReplayResponse>,
+    response_receiver: Receiver<ReplayResponse>,
+}
+
+/// A handle to the replayer. Each replayer handle has a separate channel that responses are sent on.
+/// This means multiple threads can have a handle to the replayer and results are sent back to the
+/// correct replayer handle each time.
+impl ReplayerHandle {
+    pub fn new(
+        request_sender: Sender<(Sender<ReplayResponse>, ReplayRequest)>,
+        response_sender: Sender<ReplayResponse>,
+        response_receiver: Receiver<ReplayResponse>,
+    ) -> ReplayerHandle {
+        ReplayerHandle {
+            request_sender,
+            response_sender,
+            response_receiver,
+        }
+    }
+
+    pub fn send(
+        &self,
+        request: ReplayRequest,
+    ) -> result::Result<(), SendError<(Sender<ReplayResponse>, ReplayRequest)>> {
+        self.request_sender
+            .send((self.response_sender.clone(), request))
+    }
+
+    pub fn recv_and_drain(&self) -> result::Result<Vec<ReplayResponse>, RecvError> {
+        let mut results = vec![self.response_receiver.recv()?];
+        results.extend(self.response_receiver.try_iter());
+        Ok(results)
+    }
+}
+
+impl Replayer {
```

Comment:
@apfitzge i think we can probably use this for banking stage since it's pretty generic